pkg/clusteragent/custommetrics/store.go

```diff
@@ -141,46 +142,45 @@ func (c *configMapStore) SetExternalMetricValues(added []ExternalMetricValue) er
 }
 
 // Delete deletes all metrics in the configmap that refer to any of the given object references.
-func (c *configMapStore) Delete(deleted []ObjectReference) error {
-	if c.cm == nil {
-		return fmt.Errorf("configmap not initialized")
-	}
+func (c *configMapStore) DeleteExternalMetricValues(deleted []ExternalMetricValue) error {
 	if len(deleted) == 0 {
 		return nil
 	}
-	for _, obj := range deleted {
-		// Delete all metrics from the configmap that reference this object.
-		for k := range c.cm.Data {
-			parts := strings.Split(k, keyDelimeter)
-			if len(parts) < 4 {
-				log.Debugf("Deleting malformed key %s", k)
-				delete(c.cm.Data, k)
-				continue
-			}
-			if parts[1] != obj.Namespace || parts[2] != obj.Name {
-				continue
-			}
-			delete(c.cm.Data, k)
-			log.Debugf("Deleted metric %s for HPA %s from the configmap %s", parts[3], obj.Name, c.name)
-		}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cm == nil {
+		return errNotInitialized
+	}
+	for _, m := range deleted {
+		key := externalMetricValueKeyFunc(m)
+		delete(c.cm.Data, key)
+		log.Debugf("Deleted metric %s for HPA %s/%s from the configmap %s", m.MetricName, m.HPA.Namespace, m.HPA.Name, c.name)
 	}
 	return c.updateConfigMap()
 }
 
 // ListAllExternalMetricValues returns the most up-to-date list of external metrics from the configmap.
 // Any replica can safely call this function.
 func (c *configMapStore) ListAllExternalMetricValues() ([]ExternalMetricValue, error) {
-	var metrics []ExternalMetricValue
-	if err := c.getConfigMap(); err != nil {
+	c.mu.Lock()
+	err := c.getConfigMap()
+	c.mu.Unlock()
```

Comment:
I'd prefer doing `defer c.mu.Unlock()` above and dropping the `RLock` below, this seems more readable/less error prone and the performance is not critical here

Reply:
That's true. If you'd prefer that I can do that. `ListAllExternalMetricValues` is also the function that is called the most times through the DCA so I was trying to pretend that that was a real performance improvement.

Reply:
the "possibly concurrently" hints that we should go for readability here 😁